extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-validation.service.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Validator service for the AlgebraicExpressionInput interaction.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { AnswerGroup } from
+  'domain/exploration/AnswerGroupObjectFactory';
+import { IWarning, baseInteractionValidationService } from
+  'interactions/base-interaction-validation.service';
+import { IAlgebraicExpressionInputCustomizationArgs } from
+  'extensions/interactions/customization-args-defs';
+import { Outcome } from
+  'domain/exploration/OutcomeObjectFactory';
+import { AppConstants } from 'app.constants';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AlgebraicExpressionInputValidationService {
+  constructor(
+      private baseInteractionValidationServiceInstance:
+        baseInteractionValidationService) {}
+
+  getAllWarnings(
+      stateName: string,
+      customizationArgs: IAlgebraicExpressionInputCustomizationArgs,
+      answerGroups: AnswerGroup[], defaultOutcome: Outcome): IWarning[] {
+    var warningsList = [];
+
+    warningsList = warningsList.concat(
+      this.baseInteractionValidationServiceInstance.getAllOutcomeWarnings(
+        answerGroups, defaultOutcome, stateName));
+
+    var seenIsEquivalentTo = false;
+    for (var i = 0; i < answerGroups.length; i++) {
+      var rules = answerGroups[i].rules;
+      for (var j = 0; j < rules.length; j++) {
```

Comment:
Can't there be other cases as well where rules won't get matched?
Also, what if the expressions (`inputs.x`) themselves are different between the rules in an answer group, then they may not be overruled, right?

Add more tests like these as well.

Reply:
> Can't there be other cases as well where rules won't get matched?

I don't think there are any other cases given the current rule-set. There are only 2 rules atm, so the only other case is if the MatchesExactlyWith rule precedes the IsEquivalentTo rule which is not a problem.

> Also, what if the expressions (inputs.x) themselves are different between the rules in an answer group, then they may not be overruled, right?

Oh yes, that's right. I've added a condition for that along with test cases.